plan/aggregation_push_down.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+// // Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/model"
+)
+
+func isDecomposable(fun expression.AggregationFunction) bool {
+	switch fun.GetName() {
+	case ast.AggFuncAvg, ast.AggFuncGroupConcat:
+		// TODO: support avg push down.
+		return false
+	case ast.AggFuncMax, ast.AggFuncMin, ast.AggFuncFirstRow:
+		return true
+	case ast.AggFuncSum, ast.AggFuncCount:
+		return !fun.IsDistinct()
+	default:
+		return false
+	}
+}
+
+func collectAggFuncs(agg *Aggregation, join *Join) (valid bool, leftAggFuncs, rightAggFuncs []expression.AggregationFunction) {
```

Comment:
Why collectAggFuncs returns a 'valid', does it mean here? Add comment about it.
And this is a package function, the name is general but used for aggregation push down only.